packages/astro/src/config/content.ts

```diff
@@ -0,0 +1,168 @@
+import type { Loader } from 'esbuild';
+import type { ZodObject, ZodString, ZodNumber, ZodUnion, ZodLiteral, ZodType } from 'zod';
+import { LIVE_CONTENT_TYPE, CONTENT_LAYER_TYPE } from '../content/consts.js';
+import type { LiveLoader } from '../content/loaders/types.js';
+import { AstroError, AstroErrorData, AstroUserError } from '../core/errors/index.js';
+
+export function getImporterFilename() {
+	// Find the first line in the stack trace that doesn't include 'defineCollection' or 'getImporterFilename'
+	const stackLine = new Error().stack
+		?.split('\n')
+		.find(
+			(line) =>
+				!line.includes('defineCollection') &&
+				!line.includes('getImporterFilename') &&
+				line !== 'Error',
+		);
+	if (!stackLine) {
+		return undefined;
+	}
+	// Extract the relative path from the stack line
+	const match = /\/((?:src|chunks)\/.*?):\d+:\d+/.exec(stackLine);
+
+	return match?.[1] ?? undefined;
+}
+
+// This needs to be in sync with ImageMetadata
```

Comment:
Everything below this line has been moved from `runtime.ts`. The only changes are those needed to move from a `.d.ts` to a `.ts` file